src/os/bluestore/BlueStore.cc

```diff
@@ -6518,89 +6981,39 @@ int BlueStore::read(
   dout(10) << __func__ << " " << cid << " " << oid
 	   << " 0x" << std::hex << offset << "~" << length << std::dec
 	   << " = " << r << dendl;
-  logger->tinc(l_bluestore_read_lat, ceph_clock_now() - start);
   return r;
 }
 
-// --------------------------------------------------------
-// intermediate data structures used while reading
-struct region_t {
-  uint64_t logical_offset;
-  uint64_t blob_xoffset;   //region offset within the blob
-  uint64_t length;
-  bufferlist bl;
-
-  // used later in read process
-  uint64_t front = 0;
-  uint64_t r_off = 0;
-
-  region_t(uint64_t offset, uint64_t b_offs, uint64_t len)
-    : logical_offset(offset),
-    blob_xoffset(b_offs),
-    length(len){}
-  region_t(const region_t& from)
-    : logical_offset(from.logical_offset),
-    blob_xoffset(from.blob_xoffset),
-    length(from.length){}
-
-  friend ostream& operator<<(ostream& out, const region_t& r) {
-    return out << "0x" << std::hex << r.logical_offset << ":"
-      << r.blob_xoffset << "~" << r.length << std::dec;
-  }
-};
-
-typedef list<region_t> regions2read_t;
-typedef map<BlueStore::BlobRef, regions2read_t> blobs2read_t;
-
-int BlueStore::_do_read(
-  Collection *c,
-  OnodeRef o,
-  uint64_t offset,
-  size_t length,
-  bufferlist& bl,
-  uint32_t op_flags)
+std::unique_ptr<ObjectStore::ReadTransaction>
+BlueStore::create_read_transaction(
+  const CollectionHandle& ch,
+  const ghobject_t& oid)
 {
-  FUNCTRACE(cct);
-  int r = 0;
-
-  dout(20) << __func__ << " 0x" << std::hex << offset << "~" << length
-           << " size 0x" << o->onode.size << " (" << std::dec
-           << o->onode.size << ")" << dendl;
-  bl.clear();
-
-  if (offset >= o->onode.size) {
-    return r;
-  }
-
-  // generally, don't buffer anything, unless the client explicitly requests
-  // it.
-  bool buffered = false;
-  if (op_flags & CEPH_OSD_OP_FLAG_FADVISE_WILLNEED) {
-    dout(20) << __func__ << " will do buffered read" << dendl;
-    buffered = true;
-  } else if (cct->_conf->bluestore_default_buffered_read &&
-	     (op_flags & (CEPH_OSD_OP_FLAG_FADVISE_DONTNEED |
-			  CEPH_OSD_OP_FLAG_FADVISE_NOCACHE)) == 0) {
-    dout(20) << __func__ << " defaulting to buffered read" << dendl;
-    buffered = true;
-  }
+  Collection *c = static_cast<Collection *>(ch.get());
 
-  if (offset + length > o->onode.size) {
-    length = o->onode.size - offset;
+  OnodeRef o;
+  if (c->exists) {
+    RWLock::RLocker l(c->lock);
```

Comment:
The collection is locked twice: here and in `BlueReadTransaction::read()`. Need to change that.